Refactor Controller & API

mrIncompetent opened the pull request:
**Terminology**
`Master cluster`: Initial/First/Master kubernetes cluster in a kubermatic setup. It's purpose is to serve the kubermatic dashboard and API. Also it hosts the ssh keys of a user & it currently hosts the cluster custom resources.
`Customer cluster`: A cluster, you created via the kubermatic Dashboard/Components. It mainly consists of:
- Cluster resource (CRD)
- Master Components (Deployments, ConfigMaps, Secrets)
- Nodes (Individual Nodes at a cloud-provider - managed by a separate machine/node controller)

`Seed cluster`: A kubernetes cluster which only hosts the master components of a customer cluster

**Current state**
The Cluster-Controller & all cluster resources (CR's) live inside the master cluster.
The controller will process it, and manage the master components of the cluster in the specified seed-cluster. So the controller manages multiple seed clusters
The Kubermatic API is only deployed in the master cluster and only speaks to the master cluster.

Pro:
- Single place for cluster-resources.
- Low latency for Kubermatic API calls
- Kubermatic API is just talks to a single Kubernetes cluster

Con:
- Different place for actual & desired state (namespace & cluster CR)
- Providing a master-pod logs functionality for our dashboard would mean the api needs to talk to all kubernetes clusters. (So we need to have the code for the different clusters anyway)
- We cannot use owners for the master-resources. It prevents us from reacting on a change of a customer cluster apiserver Deployment - in case we now have 0 ready apiserver pods the cluster should move into a different phase. Similar to the Node/Machine behaviour/ownership
- The controller cannot run with in-cluster config.  As we might want to open-source parts of our controller this will be beneficial.
- Single point of failure -> master-cluster

![current-state](https://user-images.githubusercontent.com/14090308/36741913-a5bd806c-1be6-11e8-93a1-0c046d0e2205.png)


**Desired State**
Cluster resources live in the seed-cluster it actually belongs to. So we have cluster CR and namespace in one seed cluster. The cluster controller also then operates in every seed-cluster.
The Kubermatic API would need to talk to each seed-cluster to fetch all clusters. To improve performance we utilize informers/listers to have a very up-to-date cache.

Pro:
- Ownership for customer cluster master resources - proper handling of actual cluster state
- Cluster only belongs into a single seed-cluster and is not split across multiple clusters
- Simple implementation for a master-pod log functionality
- Cluster-controller can run with in-cluster config

Con:
- Higher latency for Kubermatic api calls -> Informers/Listers (caching) would fix this
- Would introduce the `/dc/{dc}` parameter to all cluster related routes.
We had this initially but removed it for simplicity. Though i definitly would say the additional complexity is neglectable - at least @kgroschoff doesn't mind. 

![desired-state](https://user-images.githubusercontent.com/14090308/36743006-736db278-1be9-11e8-8c92-fd60360f03d5.png)

closes #773

mrIncompetent committed:
make controller work against exactly one cluster

mrIncompetent committed:
codegen

mrIncompetent committed:
make endpoint datacenter aware

mrIncompetent committed:
make middlewares more readable

mrIncompetent committed:
update tests

mrIncompetent committed:
fix handler tests - except admin related tests

guusvw commented on api/cmd/kubermatic-api/main.go:
```diff
@@ -27,97 +27,156 @@ import (
+		glog.Fatal(fmt.Sprintf("failed to load version yaml %q: %v", versionsFile, err))
+	}
+
+	cloudProviders := map[string]provider.CloudProvider{
```
Maybe an usecase for a registry pattern

https://github.com/golang/go/blob/go1.9/src/database/sql/sql.go#L43

mrIncompetent replied:
Thought about this as well. We actually have a function which returns this arrays. Its the parent package of the individual cloud providers. 
The issue here is, that we have a weird import order: From top-to-bottom.
And not like the go-sdk from bottom to top.

It prevents us from extracting common logic into the parent package.

Using a registry pattern would mean,
the parent package provides this registry mechanism and all cloud providers register themself during `init()` ?

guusvw replied:
Correct (y) but I already thought that we possibly can run in dependecy (circular reference) there

guusvw commented on api/cmd/kubermatic-cluster-controller/main.go:
```diff
@@ -201,45 +213,61 @@ func getEventRecorder(masterKubeClient *kubernetes.Clientset) record.EventRecord
-	seedProvider, err := seed.NewFromConfig(seedConfig)
-	if err != nil {
-		return err
+	cps := map[string]provider.CloudProvider{
```
could also handled by a registry pattern

guusvw commented on api/pkg/controller/cluster/auth.go:
```diff
@@ -19,12 +19,12 @@ import (
 )
 
-func (cc *controller) ensureRootCA(c *kubermaticv1.Cluster) error {
+func (cc *ClusterController) ensureRootCA(c *kubermaticv1.Cluster) error {
```
Do we have an usecase for exporting the `ClusterController`?

mrIncompetent replied:
I thought the rule is: "Accept interfaces return structs"
Returning a private struct doesn't make much sense, therefore i exported it.

http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/

guusvw commented on api/pkg/handler/routing.go:
```diff
@@ -11,41 +11,54 @@ import (
 )
 
+const (
+	rawToken                  = iota
```
You only need to set `iota` once

https://play.golang.org/p/EjxLdjwKPAe

https://github.com/golang/go/wiki/Iota

mrIncompetent replied:
Nice!

mrIncompetent commented:
My initial plan was actually to fix `cluster.status.phase` as we changed the behaviour of the cluster.

What i want to achieve - update the cluster phase according the actual state of the deployed manifests.
Therefore i wanted to set `metadata.OwnerReferences` on all manifests we deploy for a customer cluster. When any of those manifests change, we synchronize the `Owner` cluster and calculate the new status. Exact the same like all controllers in k8s

alvaroaleman commented on api/cmd/kubermatic-api/main.go:
```diff
@@ -27,97 +27,156 @@ import (
 func main() {
+	flag.StringVar(&listenAddress, "address", ":8080", "The listenAddress to listen on")
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to the kubeconfig.")
+	flag.StringVar(&prometheusAddr, "prometheus-address", "127.0.0.1:8085", "The Address on which the prometheus handler should be exposed")
```
Any reason why this is not on the main listener?

mrIncompetent replied:
I think we used it to not expose the metrics to the public

alvaroaleman commented on api/cmd/kubermatic-api/main.go:
```diff
@@ -27,97 +27,156 @@ import (
+	flag.StringVar(&listenAddress, "address", ":8080", "The listenAddress to listen on")
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to the kubeconfig.")
+	flag.StringVar(&prometheusAddr, "prometheus-address", "127.0.0.1:8085", "The Address on which the prometheus handler should be exposed")
+	flag.StringVar(&prometheusPath, "prometheus-path", "/metrics", "The path on the host, on which the handler is available")
```
Any reason why this is configurable?

mrIncompetent replied:
I don't really mind removing this -  i just copied it from the controller

alvaroaleman commented on api/cmd/kubermatic-api/main.go:
```diff
@@ -27,97 +27,156 @@ import (
+	datacenters, err := provider.LoadDatacentersMeta(dcFile)
 	if err != nil {
-		glog.Fatal(fmt.Printf("failed to load datacenter yaml %q: %v", *dcFile, err))
+		glog.Fatal(fmt.Printf("failed to load datacenter yaml %q: %v", dcFile, err))
```
Shouldnt this be `fmt.Sprintf`?

mrIncompetent replied:
Better would be `glog.Fatalf`

alvaroaleman commented on api/cmd/kubermatic-api/main.go:
```diff
@@ -27,97 +27,156 @@ import (
+	kubermaticMasterClient := kubermaticclientset.NewForConfigOrDie(config)
+	kubermaticMasterInformerFactory := externalversions.NewSharedInformerFactory(kubermaticMasterClient, informerResyncPeriod)
+
+	sshKeyProvider := kubernetes.NewSSHKeyProvider(kubermaticMasterClient, kubermaticMasterInformerFactory.Kubermatic().V1().UserSSHKeies().Lister())
```
`Keys` is the plural of `key`

mrIncompetent replied:
Can you be more precise? I'm not sure what you are referencing

alvaroaleman replied:
`s/UserSSHKeies/UserSSHKeys/g`

mrIncompetent replied:
That is unfortunately generated bu the Kubernetes client-generator.
We spent some time trying to change it, but its not possible at the moment.

alvaroaleman commented on api/cmd/kubermatic-api/main.go:
```diff
@@ -27,97 +27,156 @@ import (
+	}
+
+	cloudProviders := map[string]provider.CloudProvider{
+		provider.FakeCloudProvider:         fake.NewCloudProvider(),
```
Can you quickly explain why we need this? I thought we wanted to extract all cloud-specific logic into the machine-controller and have kubermatic only create crds in seed clusters.

mrIncompetent replied:
We still need to:
- Validate cloud provider specific inputs (Needs to happen in the API for UX)
- Create eventually cloud provider infrastructure (networks, security-groups, etc.)

alvaroaleman commented on api/cmd/kubermatic-api/swagger.json:
```diff
@@ -82,7 +151,7 @@
       }
     },
-    "/api/v1/cluster/{cluster}": {
+    "/api/v1/dc/{dc}/cluster/{cluster}": {
```
If we break compatibility, we shouldn't keep the version

mrIncompetent replied:
Was more a temporary thing. I wanted to introduce new v2 endpoints.
Though we need to maintain the old endpoints as well for a short period.

alvaroaleman commented on api/cmd/kubermatic-api/swagger.json:
```diff
@@ -139,7 +222,7 @@
       }
     },
-    "/api/v1/cluster/{cluster}/kubeconfig": {
+    "/api/v1/dc/{dc}/cluster/{cluster}/kubeconfig": {
```
Same as above, please change the api version if you break compatibility

mrIncompetent replied:
Was more a temporary thing. I wanted to introduce new v2 endpoints.
Though we need to maintain the old endpoints as well for a short period.

alvaroaleman commented on api/cmd/kubermatic-cluster-controller/main.go:
```diff
@@ -14,92 +14,104 @@ import (
+	flag.StringVar(&prometheusPath, "prometheus-path", "/metrics", "The path on the host, on which the handler is available")
+	flag.StringVar(&masterResources, "master-resources", "", "The path to the master resources (Required).")
+	flag.StringVar(&externalURL, "external-url", "", "The external url for the apiserver host and the the dc.(Required)")
+	flag.StringVar(&dc, "datacenter-name", "", "The name of the seed datacenter, the controller is running in. It will be used to build the absolute url for a customer cluster.")
```
Remove the coma. Maybe it would be smart to put this into a configmap if more than one controller needs it?

mrIncompetent replied:
Only one cluster-controller will run in a seed-datacenter - no need for sharing.

alvaroaleman commented on api/pkg/controller/cluster/health.go:
```diff
@@ -1,20 +1,13 @@
-	}
-
-	dep, err := informerGroup.DeploymentInformer.Lister().Deployments(ns).Get(name)
+func (cc *ClusterController) healthyDep(ns, name string, minReady int32) (bool, error) {
```
Id much prefer if this was called `healthyDeployment`, `Dep` made me first think of dependency

mrIncompetent replied:
Good point

alvaroaleman commented on api/pkg/controller/cluster/pending.go:
```diff
@@ -421,37 +382,27 @@ func (cc *controller) ensureServices(c *kubermaticv1.Cluster) error {
 		if err != nil {
 			if errors.IsNotFound(err) {
-				if _, err = client.CoreV1().ServiceAccounts(c.Status.NamespaceName).Create(generatedServiceAccount); err != nil {
+				if _, err = cc.kubeClient.CoreV1().ServiceAccounts(c.Status.NamespaceName).Create(generatedServiceAccount); err != nil {
```
Shouldn't the `generatedServiceAccount, lastApplied, err := gen(data, name, cc.masterResourcesPath)` be called here and not at the start of the loop?

mrIncompetent replied:
No, because we need the generated resource in both cases:
- When it does not exist -> Because we need to create it
- When it exists -> we need to check if we need to patch it

alvaroaleman commented on api/pkg/controller/cluster/pending.go:
```diff
@@ -473,37 +424,27 @@ func (cc *controller) ensureCheckServiceAccounts(c *kubermaticv1.Cluster) error
 		if err != nil {
 			if errors.IsNotFound(err) {
-				if _, err = client.RbacV1beta1().ClusterRoleBindings().Create(generatedClusterRoleBinding); err != nil {
+				if _, err = cc.kubeClient.RbacV1beta1().ClusterRoleBindings().Create(generatedClusterRoleBinding); err != nil {
```
Same as above, shouldn't the `clusterRoleBinding` only be created if it doesn't exist?

mrIncompetent replied:
No, because we need the generated resource in both cases:
- When it does not exist -> Because we need to create it
- When it exists -> we need to check if we need to patch it

alvaroaleman commented on api/pkg/provider/kubernetes/cluster.go:
```diff
@@ -0,0 +1,126 @@
+	// sanity checks for a fresh cluster
+	switch {
+	case user.ID == "":
+		return nil, errors.NewBadRequest("user id is required")
```
Wouldnt it be more intuitive to have one func  `VerifyNewCluster` or sth that verifies new clusters and `NewCluster` assume the cluster is valid (in all aspects that can be statically checked)? That way we could also have the API handler decide if the Error is a BadRequest or something else

mrIncompetent replied:
I agree - but the current logic is copied from the old code.
Refactoring of the validation logic should be done in a separate pr

alvaroaleman commented on api/pkg/provider/kubernetes/cluster.go:
```diff
@@ -0,0 +1,126 @@
+	return cluster, wait.Poll(10*time.Millisecond, 30*time.Second, existsInLister)
+}
+
+func (p *ClusterProvider) Cluster(user apiv1.User, name string) (*kubermaticv1.Cluster, error) {
```
I'd prefer to have this named `GetCluster` so its obvious this will never create a new one

mrIncompetent replied:
I would like to stick to the golang way of naming getters:
https://golang.org/doc/effective_go.html#Getters

alvaroaleman replied:
Ah okay, I didn't know there is a convention around that

alvaroaleman commented on api/pkg/provider/kubernetes/cluster.go:
```diff
@@ -0,0 +1,126 @@
+	return cluster, nil
+}
+
+func (p *ClusterProvider) Clusters(user apiv1.User) ([]*kubermaticv1.Cluster, error) {
```
I'd prefer to have this named `GetClusters` so its obvious this will never create a new one

mrIncompetent replied:
I would like to stick to the golang way of naming getters:
https://golang.org/doc/effective_go.html#Getters

alvaroaleman commented on api/pkg/provider/kubernetes/cluster.go:
```diff
@@ -0,0 +1,126 @@
+	return p.client.KubermaticV1().Clusters().Delete(cluster.Name, &metav1.DeleteOptions{})
+}
+
+func (p *ClusterProvider) InitiateClusterUpgrade(user apiv1.User, name, version string) (*kubermaticv1.Cluster, error) {
```
What is the difference between a `InitialClusterUpgrade` and a `ClusterUpgrade`?

mrIncompetent replied:
Its `Initiate` and not `Initial`

Though function name should probably be something like `SetVersion` or `SetClusterVersion`

alvaroaleman commented on api/pkg/provider/kubernetes/ssh.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// ClusterSSHKeys returns the ssh keys of a cluster
+func (p *SSHKeyProvider) ClusterSSHKeys(user apiv1.User, cluster string) ([]*kubermaticv1.UserSSHKey, error) {
```
please also rename to `GetClusterSSHKeys`

mrIncompetent replied:
I would like to stick to the golang way of naming getters:
https://golang.org/doc/effective_go.html#Getters

alvaroaleman reviewed (commented):
Just to clarify we're on the same page/I've understood everything, the target architecture is

1.) Have Kubermatic API  server that does CRUD on cluster crds and uses other K8s clusters that are configured as seedClusters and called DCs in the API as backing store for that
2) Have a Controller running in each seed cluster/DC that creates/updates/deletes a control plane for each cluster crd in a dedicated namespace
3) Put machine objects inside the customer cluster that are used by a machine-controller running in the seed cluster to spawn nodes

And this PR addresses the distinction of 1) and 2) and there will be another PR that addresses the distinction of 2) and 3)?

mrIncompetent commented:
@alvaroaleman 
> Just to clarify we're on the same page/I've understood everything, the target architecture is
> 
> 1.) Have Kubermatic API server that does CRUD on cluster crds and uses other K8s clusters that are > configured as seedClusters and called DCs in the API as backing store for that

Correct.

> 2) Have a Controller running in each seed cluster/DC that creates/updates/deletes a control plane for > > each cluster crd in a dedicated namespace

The CustomResources are not namespaced. But each cluster resource has a corresponding namespace which contains the control plane.

> 3) Put machine objects inside the customer cluster that are used by a machine-controller running in the > seed cluster to spawn nodes

That is something we already addressed in a dedicated PR. We added new node-endpoints. The dashboard is currently adopting them.
Though we deploy a dedicated machine-controller into each control-plane namespace.

> And this PR addresses the distinction of 1) and 2) and there will be another PR that addresses the distinction of 2) and 3)?

Mentioned above

mrIncompetent committed:
define iota once for list

mrIncompetent committed:
remove flag for prom merics path

mrIncompetent committed:
fix broken fatals in commands

mrIncompetent committed:
add backwards compatible optimistic cluster provider

mrIncompetent committed:
split v1 and v2 routes

mrIncompetent committed:
add missing healthz

mrIncompetent committed:
add v3 and fix tests

mrIncompetent committed:
add v3 to swagger swagger

mrIncompetent committed:
add missing admin functionality

mrIncompetent committed:
rebase

mrIncompetent committed:
update dep lock file

mrIncompetent committed:
remove map

mrIncompetent commented:
@alvaroaleman PTAL

alvaroaleman commented on api/pkg/controller/cluster/controller_test.go:
```diff
@@ -2,86 +2,74 @@ package cluster
-	seedcrdfake "github.com/kubermatic/kubermatic/api/pkg/crd/client/seed/clientset/versioned/fake"
-	masterinformer "github.com/kubermatic/kubermatic/api/pkg/kubernetes/informer/master"
-	seedinformer "github.com/kubermatic/kubermatic/api/pkg/kubernetes/informer/seed"
+	fake2 "github.com/kubermatic/kubermatic/api/pkg/crd/client/clientset/versioned/fake"
```
Why is it imported as fake2 and not just fake?

alvaroaleman commented on api/pkg/controller/cluster/health.go:
```diff
@@ -1,20 +1,13 @@
-	}
-
-	dep, err := informerGroup.DeploymentInformer.Lister().Deployments(ns).Get(name)
+func (cc *Controller) healthyDep(ns, name string, minReady int32) (bool, error) {
```
Please rename this to `healthyDeployment`

alvaroaleman commented on api/pkg/controller/cluster/validating.go:
```diff
@@ -32,23 +32,14 @@ func (cc *controller) validateCloudSpec(c *kubermaticv1.Cluster) error {
 	}
 	if dc.IsSeed {
-		return fmt.Errorf("given datacenter %q is not configured as a node datacenter", c.Spec.SeedDatacenterName)
+		return fmt.Errorf("specified node datacenter %q is not configured as a seed-datacenter", c.Spec.Cloud.DatacenterName)
```
Sorry for grammar nazi but in English compound nouns must not have a hyphen, that only exists in German (`s/seed-datacenter/seed datacenter/g`)

alvaroaleman commented on api/pkg/crd/client/clientset/versioned/typed/etcdoperator/v1beta2/etcdcluster.go:
```diff
@@ -1,7 +1,7 @@
 
 import (
-	scheme "github.com/kubermatic/kubermatic/api/pkg/crd/client/seed/clientset/versioned/scheme"
+	scheme "github.com/kubermatic/kubermatic/api/pkg/crd/client/clientset/versioned/scheme"
```
Why are we using named imports with the same name the import has anyways?

mrIncompetent replied:
Generated code - no chance of changing. Unless we patch the code generator in upstream kubernetes

mrIncompetent committed:
fix imports and rename healthyDep

mrIncompetent committed:
fix import

mrIncompetent commented:
@alvaroaleman all comments addressed.
I'll start updating the config.
Afterwards this PR is ready to merge.

Things to do in the next PR:
- Set actually the owner Reference in created objects

mrIncompetent committed:
update controller deployment

mrIncompetent committed:
update api deployment

mrIncompetent commented:
Config is updated.
@guusvw @alvaroaleman PTAL

Important thing to remember:
Before merging, we need to migrate some cluster resources to the seed-clusters on cloud.

mrIncompetent committed:
fix flag description

mrIncompetent committed:
fix typo

mrIncompetent committed:
fix namespace creation & fix certificate overwrites
